/*
 * Copyright(c) 2012-2018 Intel Corporation
 * SPDX-License-Identifier: BSD-3-Clause-Clear
 */
syntax = "proto3";
option cc_generic_services = true;
import "defs.proto";
import "opts.proto";
import "google/protobuf/descriptor.proto";
package octf.proto;

// Message which identifies particular command by: path, interface, and command
// key (long or short)
message CliCommandId {
    // CLI long or short command switch
    string commandKey = 1;
}

// Description of command
message CliCommandDesc {
    // Path of node which provides given interface
    NodePath path = 1;

    // ID of interface which implements given method
    InterfaceId interface = 2;

    // Method ID within given interface
    int32 methodId = 3;

    // CLI commands option (like: short key, long key, description for help)
    OptsCommand cmdOps = 4;

    // Descriptors of input and output types for this command
    google.protobuf.FileDescriptorSet inputOutputDescription = 5;

    // Full name of input message type for this command
    string inputMessageTypeName = 6;

    // Full name of output message type for this command
    string outputMessageTypeName = 7;
}

// Set describing available commands
message CliCommandSetDesc {
    // Array of commands
    repeated CliCommandDesc command = 1;
}

// Auto CLI Interface definitions
service InterfaceCli {
    // Get all CLI commands description
    rpc getCliCommandSetDescription(Void) returns (CliCommandSetDesc);

    // Get particular command description
    rpc getCliCommandDescription(CliCommandId) returns (CliCommandDesc);
}
